### ISSUE TITLE
Make Problems point at nodes

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#955]
Make it possible to point a `Problem` at (from least specific to most specific):
- nothing
- `Range` (old situation)
- `TokenRange` (current situation)
- `Node` (interesting new situation)

**Comment by matozoid** [comment]
This was thought up while adding recovery in the grammar, since we are explicitly creating a node that has a clear problem: it didn't parse. It would be nice if the `Problem` could directly point at that node, well, that would be very clean.

**Comment by matozoid** [comment]
Of course, if we know the `TokenRange`, we know the `Range`, and if we know the `Node` we know the `TokenRange`.

**Comment by ryan-beckett** [comment]
@matozoid: I want to remove

`Problem.Problem(String message, TokenRange location, Throwable cause)`

and replace it with:

`public Problem(String message, Node node, Throwable cause)`

*Issues*

1. No other JavaParser code uses the `TokenRange` parameter, so we'd be fine there.
2. Would client code break?

If neither of these are potential issues, are we good to go?

**Comment by ryan-beckett** [comment]
Looking at `GeneratedJavaParser`, I really don't see how it's possible to point a `Problem` to a `Node` that may not even exist. If I'm not mistaken, if a `ParseException` is thrown in `JavaParser.parse`, the AST created in `GeneratedJavaParser` is thrown away and the token information where parsing failed is printed into the exception message.

**Comment by ryan-beckett** [comment]
@matozoid: Just to clarify, in `java.jj`, when an exception is thrown the only node types ever returned are `UnparsableStmt`, `CompilationUnit`, and `BlockStmt`. Are these the nodes that you'd like `Problem` to point to (which doesn't add any value IMO), or are did you want to be a lot more specific than that. For instance, if the parser fails on trying to match the tokens for a `LabeledStatement`, you want the error node of `Problem` to be a `LabeledStatement`. If it's the latter, this seems like a very difficult improvement.

**Comment by matozoid** [comment]
@ryan-beckett - I appreciate you diving into error recovery, it's a difficult problem where work only just started. In case you haven't seen it, here's a blog I wrote with some background: https://matozoid.github.io/2017/06/11/parse-error-recovery.html

So, those three that are returned are only the start. I'm not yet sure how far we can go with recovery, but here's a quick list I made by going through the grammar (java.jj) https://github.com/javaparser/javaparser/issues?q=is%3Aissue+is%3Aopen+label%3A%22error+recovery%22

`LabeledStmt` is a good example: if you were to add more specific recovery, which seems like a good idea since it is easy to search for ":" when things go wrong. Then you could create a very local `Problem` pointing at that node.

The eventual goal would be to match the power of popular IDE's: make a typo somewhere, and you get an indication of what's wrong that's as specific as possible. This goal is unreachable since we don't have the commercial teams they have, but we can see how far we can get :-)

If you think it's a bit too hard then you're free to pick up something else of course ;-)

**Comment by ryan-beckett** [comment]
@matozoid: Thanks for the clarification. At least I have a better idea of what's going on. Yeah, I'm going to work on something where I need as little knowledge on the inner workings as possible.

**Comment by ryan-beckett** [comment]
@matozoid: Danny, what's a good way to contact you about a question unrelated to JP? I don't think Tweeter is proper channel either. Would **hexagonaal@gmail.com** work?

**Comment by ryan-beckett** [comment]
@matozoid: Is there a way I can contact you for a discussion unrelated to JP? How about **hexagonaal@gmail.com**? I can't DM you on Twitter, since you don't follow me.

**Comment by matozoid** [comment]
@ryan-beckett I'd say Gitter, a lot of JP discussion is there and you can also send private messages.

**Comment by matozoid** [comment]
https://gitter.im/javaparser/javaparser
